src/cryptoadvance/specter/user.py

```diff
@@ -36,69 +44,146 @@ def verify_password(stored_password, provided_password):
-        """pass a json or a plain-password here"""
-        try:
-            if value.get("salt") and value.get("pwdhash"):
-                self._password = value
-        except:
-            salted_hashed_password = hash_password(value)
-            self._password = salted_hashed_password
+        # use the Scrypt KDF to get a private key from the password
+        private_key = hashlib.scrypt(
+            plaintext_password.encode(), salt=salt, n=2 ** 14, r=8, p=1, dklen=32
+        )
 
-    @classmethod
-    def from_json(cls, user_dict, specter):
-        # TODO: Unify admin in backwards compatible way
-        try:
-            if not user_dict["is_admin"]:
-                return cls(
-                    user_dict["id"],
-                    user_dict["username"],
-                    user_dict["password"],
-                    user_dict["config"],
-                    specter,
-                )
-            else:
-                return cls(
-                    user_dict["id"],
-                    user_dict["username"],
-                    user_dict["password"],
-                    {},
-                    specter,
-                    is_admin=True,
+        # create cipher config
+        cipher_config = AES.new(private_key, AES.MODE_GCM)
+
+        # return a dictionary with the encrypted text
+        cipher_text, tag = cipher_config.encrypt_and_digest(self.plaintext_user_secret)
+        self.encrypted_user_secret = {
+            "cipher_text": b64encode(cipher_text).decode("utf-8"),
+            "salt": b64encode(salt).decode("utf-8"),
+            "nonce": b64encode(cipher_config.nonce).decode("utf-8"),
+            "tag": b64encode(tag).decode("utf-8"),
+        }
+
+    def decrypt_user_secret(self, plaintext_password):
+        # See: https://qvault.io/cryptography/aes-256-cipher-python-cryptography-examples/
+        if not self.encrypted_user_secret:
+            self._generate_user_secret(plaintext_password)
+
+        # decode the dictionary entries from base64
+        salt = b64decode(self.encrypted_user_secret["salt"])
+        cipher_text = b64decode(self.encrypted_user_secret["cipher_text"])
+        nonce = b64decode(self.encrypted_user_secret["nonce"])
+        tag = b64decode(self.encrypted_user_secret["tag"])
+
+        # generate the private key from the password and salt
+        private_key = hashlib.scrypt(
+            plaintext_password.encode(), salt=salt, n=2 ** 14, r=8, p=1, dklen=32
+        )
+
+        # create the cipher config
+        cipher = AES.new(private_key, AES.MODE_GCM, nonce=nonce)
+
+        # decrypt the cipher text
+        self.plaintext_user_secret = cipher.decrypt_and_verify(cipher_text, tag)
+
+    def _generate_user_secret(self, plaintext_password):
+        # Encryption using the user_secret uses a Fernet key. But the Fernet
+        #   key itself will be encrypted with the user's password.
+        self.plaintext_user_secret = Fernet.generate_key()
```

Comment:
I'm a bit confused. Shouldn't that be done just initially? If the user sets a new password, the `self.plaintext_user_secret` should be the same, no? Here, it's created anew every time this method get called (e.g. in `set_password`).

Reply:
Yes, I think you're right. The existing `plaintext_user_secret` needs to be re-encrypted on password change but regenerating would render the downstream encrypted data orphaned.